### PR TITLE
WIP tabix

### PR DESCRIPTION
**Pull request by h-2**
This PR adds Tabix supports and indexed VCF reading.

All Tabix code is currently detail and will probably stay there for now. 

Some preliminary "benchmarks":

|                            |  bio w/o index | bio w/ index | bcftools w/ index |
|------------------------|-------------------:|----------------:|----------------------:|
| filesystem inputs |      628,748    |    18,557      |      13,051           |
| time                     |          53.1s   |    1.7s           | 3.2s                    |

Due to architectural problems, I don't think we can ever get the IOPS to be as low as with htslib. Please see my comments in the PR. In practice, the results seem to still be OK, but this is just one example where I tried a region very far to the end of a 300MB compressed VCF. We definitely need to do more testing.

TODO
- [x] stand-alone tests for tabix input
- [ ] do we want to add files to our unit tests that result in multiple BGZF blocks? Probably good idea, but also big.
- [ ] discuss indexing strategy with @smehringer and @joshuak94

### CODE REVIEW

File: include/bio/var_io/reader.hpp [hunk 1]
```diff
@@ -126,6 +128,121 @@ class reader : public reader_base<reader<option_args_t...>, reader_options<optio
...
+                index.read(index_path);
+                std::vector<std::pair<uint64_t, uint64_t>> chunks = index.reg2chunks(options.region);
+
+#if 1 //linear implementation
+                /* IMPLEMENTATION NOTE
+                 * We are currently doing a simplified indexed access where we do not process all
+                 * possible chunks but just do a linear scan from the beginning of the first overlapping chunk.
+                 * This is definitely worse than what htslib is doing, but still very good for the tests performed.
+                 * I doubt that we can ever reach htslib's performance fullty while using C++ iostreams.
+                 */
```
Comment (h-2) [inline include/bio/var_io/reader.hpp:154]: What are your thoughts on this, @joshuak94 and @smehringer ?

**Comment by Irallia** [comment]
Do you know about the [BAMIntervalTree](https://github.com/joshuak94/BAMIntervalTree) created by @joshuak94 for indexing BAM files in another way?

**Comment by h-2** [comment]
> Do you know about the [BAMIntervalTree](https://github.com/joshuak94/BAMIntervalTree) created by @joshuak94 for indexing BAM files in another way?

Thanks for the pointer! I talked to him about. But even if we want to support that, we also need to be able to handle the regular indexes.

**Comment by h-2** [comment]
It doesn't yet contain all the tests I would have liked, but this is as much as I can currently do for this feature.

**Pull request merged**